clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -117,15 +117,21 @@ public ByteBuffer allocate(int size, long maxTimeToBlock) throws InterruptedExce
                 int accumulated = 0;
                 ByteBuffer buffer = null;
                 Condition moreMemory = this.lock.newCondition();
+                long remainingTimeToBlockNs = TimeUnit.MILLISECONDS.toNanos(maxTimeToBlockMs);
                 this.waiters.addLast(moreMemory);
                 // loop over and over until we have a buffer or have reserved
                 // enough memory to allocate one
                 while (accumulated < size) {
-                    long startWait = time.nanoseconds();
-                    if (!moreMemory.await(maxTimeToBlock, TimeUnit.MILLISECONDS))
-                        throw new TimeoutException("Failed to allocate memory within the configured max blocking time");
-                    long endWait = time.nanoseconds();
-                    this.waitTime.record(endWait - startWait, time.milliseconds());
+                    long startWaitNs = time.nanoseconds();
+                    boolean waitingTimeElapsed = !moreMemory.await(remainingTimeToBlockNs, TimeUnit.NANOSECONDS);
+                    long endWaitNs = time.nanoseconds();
+                    long timeNs = Math.max(0L, endWaitNs - startWaitNs);
+                    this.waitTime.record(timeNs, TimeUnit.NANOSECONDS.toMillis(endWaitNs));
```

Comment:
@ijuma Do you mean changing `this.waitTime.record(timeNs, TimeUnit.NANOSECONDS.toMillis(endWaitNs))` to `this.waitTime.record(timeNs, time.milliseconds())`? I think the only difference between the two is the time to do `long timeNs = Math.max(0L, endWaitNs - startWaitNs) `, which shouldn't take much time. Do we need this extra accuracy at the cost of one extra system call `time.millionseconds()`?